### ISSUE TITLE
[NEW] Valkey Modules Versioning & Bundling

### ISSUE THREAD

**Issue by hpatro** [madolson/placeholderkv#408]
Valkey supports dynamic loading of modules, which expands its capabilities by allowing users to add functionality beyond the core data structures at runtime. This feature enables users to enhance the core engine with custom modules developed independently. Bundling popular modules such as Bloom filters, JSON, Search, Timeseries, etc., along with the core engine enables adoption of Valkey for users seeking these features and also simplifies transitioning from Redis to Valkey. 

Valkey can pursue one of the following options regarding bundling of Modules for each release version:

1. **Generate two binaries - Valkey Core (only the core) and Valkey Plus (core + modules)**: This option maintains a lightweight core binary and allows users to choose which modules to load. Additionally, a bundled binary includes popular and Valkey organization-supported modules.
2. **Generate a single binary with all Valkey-supported modules** - This option involves creating a single binary of the core along with all Valkey organization-supported modules. This approach simplifies the release process by requiring only one binary.
3. **Generate Valkey core binary and each modules binary independently** - This option allows different release cycle for the core and modules. The release process can be decoupled and can be performed independently maintainers of each project.

Personally, I prefer option 1 as it provides flexibility to users in choosing modules according to their needs as well as avoids the complexity of loading modules separately if they need any of the module based features.

------------------------------------------------------------------------------------------------------------

Summary from [discussion](https://github.com/valkey-io/valkey/issues/408#issuecomment-2696035333) on March 03 2025 Valkey Team meeting:

* Versioning
    *  **Valkey core** - It will continue with it's existing version strategy. No changes in it.
    * **Modules** - will have their independent version release mechanism and will start off from 1.0.0. They will also follow [semantic versioning guidelines](https://semver.org/). Each module within the repository will define their compatibility with Valkey core, further explained in https://github.com/valkey-io/valkey/issues/408#issuecomment-2696035333
    * **Valkey with Modules** - This will be a bundle of Valkey core + GA Modules. This will also follow semantic versioning. The major and minor version of the bundled release will be tied up to the Valkey core's major/minor version. The patch version will be bumped for any latest patches picked for core/modules.

* Artifact Release
    *  **Valkey core** - It will continue to be released as is. The binaries will be available on the the valkey.io website and  containers via docker images.
    *  **Valkey with Modules** - This will be the new artifact release going forward and will contain the core with all the modules which have GA'd. 

Ref: https://github.com/valkey-io/valkey/issues/407

**Comment by PingXie** [comment]
Does option 1 require a change to the module build process? We currently build each module into its own .so file.

Does option 2 affect the distribution only? or it implies that all modules distributed along with the core will be loaded automatically?

**Comment by natoscott** [comment]
I think there's an option #4 too which is an extension to #3:   Valkey as it is now, in a container, with some blessed group of modules installed, and suitable valkey.conf settings to ensure those modules are automatically loaded at startup.  That would be one definition of "bundling" anyway, that doesn't require any new code, makefile or other linkage changes to Valkey itself.

This option provides the benefits of all three of the listed options, I think...?  And its what all the cloud vendors need (a container) - and can form the basis of future work on a Valkey Kubernetes operator.

**Comment by hwware** [comment]
> Does option 1 require a change to the module build process? We currently build each module into its own .so file.
> 
> Does option 2 affect the distribution only? or it implies that all modules distributed along with the core will be loaded automatically?

IMO, I do not like option 2, because module is similar to a plugin. With more and more modules appear, the binary file could be very big. We should allow clients to choose which module(s) they want to use.  There are thousand of plugins for eclipse and VS code 

I prefer option 3. option 1 is a a little bit weird.

**Comment by madolson** [comment]
Definitely don't like 2. We would like to have a lean distribution of Valkey. I'm also leaning towards 3 over 1.

> Does option 1 require a change to the module build process? We currently build each module into its own .so file.

It would yeah, we would probably have a separate build repository that builds everything together with special flags. I will say it make it much more difficult for end users to build, as you'll probably need to check out all the sub-modules and build them all individually. We've basically defined three distribution channels: direct downloads, containers, and through linux distros. I think this approach only really helps the first and third since you have everything built, but I have a suspicion that both 1 and 3 would be okay with downloading the modules separately and having to do a little assembly on their end.

> Valkey as it is now, in a container, with some blessed group of modules installed, and suitable valkey.conf settings to ensure those modules are automatically loaded at startup. 

I think this makes a lot of sense. This is the easiest way to "download" and try out the functionality.

**Comment by hpatro** [comment]
> > Does option 1 require a change to the module build process? We currently build each module into its own .so file.
> > Does option 2 affect the distribution only? or it implies that all modules distributed along with the core will be loaded automatically?
> 
> IMO, I do not like option 2, because module is similar to a plugin. With more and more modules appear, the binary file could be very big. We should allow clients to choose which module(s) they want to use. There are thousand of plugins for eclipse and VS code
> 
> I prefer option 3. option 1 is a a little bit weird.

Option 3 is a subset of option 1, the only issue I see with it is some additional management operation needs to be performed by the admin to vet/load the module(s) on the node(s). 
We can start off with option 3 and see if any user wants this bundling (option 1) and support it in the future.

**Comment by daniel-house** [comment]
How much risk is there that putting multiple modules written by various authors into a single binary will cause identifier collisions leading to undefined behavior?

hmmmm, this is making me think about the difference between the valkey binary and the shared libraries.  Won't options 1 and 2 require code changes to allow non-loading of the modules that the user doesn't want?  How would we even do that, given that the existing code searches the library for a function named XXXModule_OnLoad, and calls it for each module?

What am I missing?

**Comment by murphyjacob4** [comment]
> I have a suspicion that both 1 and 3 would be okay with downloading the modules separately and having to do a little assembly on their end.

For iterative features, I think this makes sense (Option 3).

But if we ever wanted to take functionality out of the core and put it into a module - it would probably break a lot of users who now would need to do `loadmodule` to get that functionality (e.g. I've heard some discussion of moving it to a module instead of a core engine feature). This is where I think something like Option 1 could be appealing - we can make decisions about spinning off features into modules without necessarily changing the functionality of the build artifact we ship to customers.

If we did go with Option 1, I think we should be selective about what gets automatically bundled, it should only be widely-used production-ready modules.

**Comment by stockholmux** [comment]
I'd like to avoid bundling modules, so option 3 (or maybe even looser).

I don't like option 1. It picks winners. Say, I create a module that uppercases a string (silly example). It gets accepted in Valkey Plus. Then someone else comes along and creates a better uppercase module with a different API. Then the TSC has to make a gross decision: have two, incompatible uppercase modules (yuck), drop one for the other (breakage), or let the better module wallow in disuse because it's not in Valkey Plus (yuck). 

I don't like option 2 and it troubles me about what Valkey even is. It has all the problems of option 1 AND why even have a code and modules then? To a user, they'll see commands and not differentiate between core commands and module commands, meaning the project will probably need to treat them the same way for versioning, maintenance, etc. 

Frankly, I'd love to keep modules outside of Valkey. Maybe create a registry and an easy way to install modules from the registry.

Any way you cut it, the user should be in control and the bundling should minimize situation where the bundle picks winners.

**Comment by daniel-house** [comment]
> But if we ever wanted to take functionality out of the core and put it into a module - it would probably break a lot of users who now would need to do `loadmodule` to get that functionality (e.g. I've heard some discussion of moving it to a module instead of a core engine feature).

This seems like an important observation.  There are many references to moving core features into modules.  A very significant recent suggestion is to move the consensus algorithm into a module in cluster V2.  Presumably this really means creating module-API functions that allow the default implementation of a feature to be overridden by means of the new module-API functions.  I'd very much like to hear other peoples thoughts about what "move it into a module" means.

**Comment by hpatro** [comment]
> If we did go with Option 1, I think we should be selective about what gets automatically bundled, it should only be widely-used production-ready modules.

@murphyjacob4 that's the idea with option 1, I believe the TSC/maintainers are the best set of folks to take the decision on behalf of the users and reduce some of the administrator pain. And if in future if there is a better alternative (spec/performance/feature/memory usage) a drop in replacement can be performed without impacting the users.

**Comment by dmitrypol** [comment]
I am also for option3. I like the idea of official modules ( and user developed modules).  But I think it's better to have separate releases for Valkey core vs each module.  Each module will have its own feature set, engineers and hence timelines.  

eventually if a module becomes super popular with everyone using it then we could move that code into the Valkey code but then it's no longer a module but part of core.

**Comment by PingXie** [comment]
>I'd very much like to hear other peoples thoughts about what "move it into a module" means.

I think we are a bit off topic since the proposal here is about the "true" modules, for lack of a better word, such as JSON/Bloom filter/etc. I am partially guilty for this, given that I used "modules" a lot in the cluster v2 discussion.

@daniel-house, agreed the "modularization" idea that we were discussing in the cluster V2 thread need further clarification/deep-dive. At least from my end, I have been using "modules" in a very loose way in the context of the cluster V2 discussion. Let me expand my thoughts a bit more to help distinguish the two types of "modules" so we can focus back on the "true" modules on this thread and continue the "modularization" discussion separately. 

First of all, I am fully aware of the operational convenience that we only have to deal with a single binary today. Not saying we should never break away from it but I think there is an extremely high bar that should be met before we start introducing a collection of binaries. On the other end, I don't like that there is no clear layering nor strict contracts between the logical components in the engine, such as clustering, persistence, and replication, etc. The recent refactoring of cluster.c is not helping either IMO. There needs to be a clean contract/mechanism that allows us to abstract away the low level implementation from the rest of the engine. This mechanism could be based on the existing module APIs or a further extension of it; or it could be something totally different. The keyword IMO is "abstraction" and this is what I had in mind when I wrote "modules" or "modularization" in the cluster V2 thread. And to be clear, I am not advocating we create separate binaries for cluster (v1 or v2). There should still be one `valkey-server` after the `modularization` work. I don't have details about what this new abstraction layer would look like right now. I will be trying some random ideas next.

**Comment by ranshid** [comment]
Although late for this discussion, would also like to ask some questions about modules SDK. Currently we keep only the [rust module SDK](https://github.com/valkey-io/valkeymodule-rs) under our repository. In case of option 1, how will we import the correct SDK? I also think that our module "certification" process should require it to be able to build with our supported SDKs.

**Comment by madolson** [comment]
Next step from core team meeting:
1. Ask the modules to determine a 1.0.0 major release, and have them release independently for that. 
2. We will follow up about how we want to release a distribution of the modules. @madolson Says she is working on an RFC that they will post.
3. We have not decided how future module versions will declare support for core versions. (owner TBD)
4. We will define a process for modules to add their commands and documentation on to the official website. (owner TBD)

**Comment by PingXie** [comment]
Thinking about this some more, I agree it makes more sense at this point to release the modules independently rather than bundling them with the core engine. Bundling is a significant and long-term commitment, and we should ensure we have a well-defined and robust deprecation process in place before taking that step. Starting with independent releases under a clear process allows us to build confidence and flexibility as we move forward. 

I am aligned with https://github.com/valkey-io/valkey/issues/408#issuecomment-2546223718.

**Comment by yisraelU** [comment]
As both latecomer and bystander , however a Valkey user , I like Option 3 . I dont like the coupling created with the other 2 options . Also I suspect many users might be using some sort of container technology like Docker , which provides another layer with which to create desirable and easily configurable combos of core and modules to suit one likings.

**Comment by madolson** [comment]
### Core team leaning towards
1. We will release a new distribution, "Valkey-bundle". When downloaded it will be an archive of modules. When pulled as a container, it will include all the modules loaded.
2. The versioning of the "Valkey-bundle" will have the same major and minor as Valkey-core.
3. The bundle will follow semantic versioning, each time a module or the core releases a patch, the bundle will bump the patch version.
4. Modules will declare their minimum supported version, so if someone is rolling their own.
5. We'll target launched Valkey module distribution 8.1 at the OSS Summit NA, which is in June.

### Pending open questions
1. What will the documentation story?
2. What will we name the large of the larger bundling strategy?

**Comment by sarthakaggarwal97** [comment]
# Valkey Versioning and Release Strategy 

In our approach, the core platform and its modules evolve independently while strictly adhering to Semantic Versioning (SemVer). The key proposal includes modules maintaining their own version numbers, defining compatibility for entire major versions of Valkey, and also to define a version and release strategy for bundling of core and module for distributions.

## Motivation

Valkey’s ecosystem comprises a core platform and a suite of modules that extend its functionality. Currently, the core platform is scheduled for release as version **8.1.0**, whereas modules begin at **1.0.0**. Although both components adhere to **SemVer**, their independent evolution has not been formally orchestrated. This lack of a unified strategy can lead to integration challenges, increased backporting efforts, and ambiguous compatibility expectations among users.

### Objectives

The primary objectives of this strategy are to:

* **Decouple the Versioning Trajectories:** Allow the core platform and its modules to evolve independently, ensuring that changes in one do not impose unnecessary version updates in the other.
* **Maintain Semantic Integrity:** Rigorously adhere to SemVer rules so that any breaking changes are clearly and unambiguously communicated.
* **Guarantee Compatibility:** Ensure that minor and patch releases of both the core and modules remain compatible across the ecosystem.

## Versioning and Compatibility Model

### Independent Versioning Trajectories [Preferred]

Drawing upon principles from SemVer, modular systems and loosely coupled architectures, Valkey is envisioned as a collection of distinct, interacting modules. The core platform provides the foundational services, while modules augment functionality without altering the core’s invariants. A key innovation in our model is the use of a **compatibility descriptor**—a metadata field embedded in each module that declares the range of core versions with which it is compatible.

In our decoupled model, a module can release a major version independently of a Valkey (core) major version change. In this approach, each module follows its own SemVer trajectory—if a module introduces breaking changes or fundamentally alters its API, it must bump its major version, even if the core platform remains unchanged. The key is that the compatibility descriptor (e.g., `^8.0.0`) in the module metadata is updated accordingly to reflect which core versions it supports

* **Core Platform Versioning:**
    Valkey’s core platform will be released as version **8.1.0**, with subsequent releases following the standard SemVer rules to reflect genuine changes in functionality or compatibility.
* **Module Versioning:**
    Modules will commence at version **1.0.0** and evolve independently. Their version increments will reflect internal modifications, enhancements, and bug fixes without being directly coupled to core platform version changes.

### Compatibility Declaration

Each module will include a *compatibility descriptor* within its metadata (e.g., in a properties or manifest file) that specifies a SemVer range representing the supported core platform versions. For example, a module might declare compatibility as `^8.0.0`, indicating that it is guaranteed to function correctly with any core platform release within the 8.x series.

#### Compatibility Matrix Illustration

|Module Version	|Declared Core Compatibility	|Implication	|
|---	|---	|---	|
|1.0.0	|`^8.1.0`	|Compatible with all 8.x releases; no breaking core API changes expected.	|
|1.1.0	|`^8.1.0`	|Minor enhancements; continues to work seamlessly with core 8.x.	|
|2.0.0	|`^9.0.0`	|Represents a breaking change in the module if necessitated by a revised core API.	|

With every Valkey release, we will publish the set of supported and tested versions of modules.

### Handling Major Version Transitions in the Core

A central question is whether a major version change in the core platform should necessitate a corresponding major version change in modules.

* **Scenario 1: Non-Breaking Core Transition**
    If the core advances from 8.x to 9.x while maintaining backward compatibility for module interfaces, modules may simply update their compatibility descriptors (e.g., from `^8.0.0`) without a major version increment.
* **Scenario 2: Breaking Core Changes**
    Should the core major version change introduce incompatible modifications in the module APIs, modules must release a new major version (e.g., upgrading from 1.x to 2.0.0) to signal these changes as required by SemVer.

Thus, a module’s major version increment is contingent upon whether its operational contract with the core is altered, not merely as a function of the core’s version progression.

**Sample Compatibility Descriptor File.**

These fields are proposed for initial descriptor file. We can add more fields as deemed necessary. Although, we should mandate some of the fields to be present in the descriptor when trying to load a module, so Valkey core can make some of the changes.

```
# ValkeyJSON Module Descriptor

module.name=ValkeyJSON
module.version=2.0.0
module.description=Provides JSON parsing and serialization support for Valkey core services.

# Declares compatibility with all Valkey core versions in the 8.x series
module.compatibility=^8.0.0
module.releaseDate=2025-02-19
```

## Alternative Approach: Synchronizing Module and Core Major Versions

An alternative strategy under consideration is to synchronize the major version numbers of Valkey modules with the core platform. In this approach, a module's major version would mirror the core's major version (e.g., core 8.x would mandate modules to be labeled as 8.y.z). While the modules could still independently manage their minor and patch numbers, the major version would be tied directly to the core.

We can still work with the four part versioning system, where any module patch / minor version updates will be accounted for in the release notes and through the bump fourth part (by the number of updates).

#### **Pros:**

* **Consistency and Predictability:**
    Users could more easily infer compatibility; if a module’s major version matches the core’s, it suggests that the module is designed to work with that core series. This creates a clear, unified versioning ecosystem.
* **Simplified Compatibility Communication:**
    Aligning major versions removes ambiguity about whether a module is intended for use with the current core release, potentially reducing user error and support overhead. We would not need compatibility matrix or descriptors to specify the valid combination of modules and core Valkey.

**Cons:**

* **Reduced Flexibility:**
    Modules may evolve at a different pace than the core platform. For instance, a module might not require a breaking change even when the core does, yet synchronizing major versions would force an unnecessary major version bump. 
* **Increased Release Overhead:**
    Mandating a major version update for every core release may require module maintainers to perform additional release activities, even when their code remains functionally unchanged. 


## Distribution Strategies

### Bundled Distribution [Preferred]

**Composite Releases:**
In bundled distribution, modules are packaged alongside the core platform. A composite versioning scheme (e.g., a three part version (following SemVer) such as `8.1.X`) is used, where the first two numbers reflect the core platform major/minor version and the third indicates patch releases for any update on core or modules. Each distribution release will maintain its own release notes, documenting which components were updated and linking to their respective detailed release notes.

* **Two-Tiered Distribution Model
    **Considering the varied needs of our user base, we propose offering two types of distributions:
    * **Valkey Core:**
        This distribution releases core Valkey without any of its modules.
    * **Valkey with extensions:
        **Contains all available modules, including experimental or less-validated versions. Caters to advanced users or developers who require access to the full spectrum of functionality and are comfortable managing potential integration issues.

Bundled releases ensure that core and module changes are integrated and tested together, providing a cohesive update experience for end users.

### Standalone Module Distribution

* **Independent Releases:**
    Alternatively, modules may be released as standalone artifacts. This model allows for rapid deployment of module updates without necessitating a full core upgrade.
* **Clear Artifact Identification:**
    Standalone artifacts will be labeled with both the module version and the supported core version range, ensuring transparency in compatibility. Release artifacts will be generated with clear naming schemes such as:
    `valkey-module-1.2.3-Core_8.x.y.zip`, which succinctly communicates the module version and its compatible core range.

* * *

## Release Ownership [TBD]

It is proposed that a module releasing the patch update, to coordinate the release of the distribution across the bundle. With this, we can specify the set of instructions (automate wherever possible) to release the latest stable distribution with all the updated components.
* * *

**Comment by hpatro** [comment]
This was approved long back in March 2025 and the strategy has been implemented for Valkey core and Valkey bundle. Just added the major-decision-approved tag now.
